### PR TITLE
linkers: Fix RSP syntax for linkers invoked with clang

### PR DESCRIPTION
**Pull request by kasper93**
Fixes: #8981
Fixes: 2be074b1d445fcd30535bcf7518f8ce0738bcbf3

### CODE REVIEW

File: mesonbuild/linkers/detect.py [hunk 1]
```diff
@@ -92,7 +93,10 @@ def guess_win_linker(env: 'Environment', compiler: T.List[str], comp_class: T.Ty
         elif not invoked_directly:
             return ClangClDynamicLinker(
                 for_machine, override, exelist=compiler, prefix=comp_class.LINKER_PREFIX,
-                version=search_version(o), direct=False, machine=None)
+                version=search_version(o), direct=False, machine=None,
+                rsp_syntax=RSPFileSyntax.MSVC
+                            if (invoked_directly or comp_class.get_argument_syntax() == 'msvc')
```
Comment (xclaesse) [inline mesonbuild/linkers/detect.py:98]: That code is inside the block `elif not invoked_directly:` so the condition can be simplified to `if comp_class.get_argument_syntax() == 'msvc'` AFAIK.

Comment (xclaesse) [reply mesonbuild/linkers/detect.py:98]: Actually, I don't even know what `invoked_directly` really means here? I'm not expert in that part of the code.

File: mesonbuild/linkers/linkers.py [hunk 2]
```diff
@@ -1235,11 +1235,13 @@ def _apply_prefix(self, arg: T.Union[str, T.List[str]]) -> T.List[str]: ...
 
     def __init__(self, exelist: T.List[str], for_machine: mesonlib.MachineChoice,
                  prefix_arg: T.Union[str, T.List[str]], always_args: T.List[str], *,
-                 version: str = 'unknown version', direct: bool = True, machine: str = 'x86'):
+                 version: str = 'unknown version', direct: bool = True, machine: str = 'x86',
+                 rsp_syntax: RSPFileSyntax = RSPFileSyntax.MSVC):
```
Comment (xclaesse) [inline mesonbuild/linkers/linkers.py:1239]: Shouldn't the default be GCC? It seems you only give a value to msvc/cl linkers.

**Comment by xclaesse** [comment]
(Ignore my previous comments, I have a better understanding now)

What feels wrong is `Compiler.rsp_file_syntax()` returns `self.linker.rsp_file_syntax()`. See https://github.com/mesonbuild/meson/blob/master/mesonbuild/compilers/compilers.py#L632 and https://github.com/mesonbuild/meson/blob/master/mesonbuild/compilers/compilers.py#L1326. If linker and compiler can have different format, ninja should get the value for compiler and linker separately when they are direct. See `generate_dynamic_link_rules()` it does [`options = self._rsp_options(compiler)`](https://github.com/mesonbuild/meson/blob/95b03f793028968a4a6df7fe25358fd8c5d39a0c/mesonbuild/backend/ninjabackend.py#LL2298C17-L2298C17). Shouldn't that be `options = self._rsp_options(compiler.linker if compiler.linker else compiler)` or something like that?

**Comment by kasper93** [comment]
> What feels wrong is Compiler.rsp_file_syntax() returns self.linker.rsp_file_syntax().

I need to dig into it to understand, but looks like current logic considers only the linker input to be "rspable". I guess linker commands are commonly the longest one. I will look into it when I have time.

**Comment by xclaesse** [comment]
> I guess linker commands are commonly the longest one.

That's the most likely indeed, but ninja backends calls `self._rsp_options(compiler)` in `generate_compile_rule_for()`, so that gets used to compile single objects as well, as far as I understand.

**Comment by kasper93** [comment]
@xclaesse: I've updated the PR. I think is all I can do. I never intended to start refactoring half of codebase, but only fix an issue that prevents us from using meson with MSVC clang.

There is `get_argument_syntax()` and `rsp_file_syntax()` which ultimately means exactly the same thing, but refactoring all this requires in-depth look into every compiler/linker definition, which I'm not willing to do at this point. I fact 1.2.0 meson is not working at all for me, so it is even harder to work on it.

**Comment by kasper93** [comment]
Could we merge this as a regression fix and delegate any improvement to future PRs?

File: mesonbuild/compilers/asm.py [hunk 3]
```diff
@@ -157,7 +157,8 @@ class MasmCompiler(Compiler):
     def get_compile_only_args(self) -> T.List[str]:
         return ['/c']
 
-    def get_argument_syntax(self) -> str:
+    @staticmethod
```
Comment (bruchar1) [inline mesonbuild/compilers/asm.py:160]: This addition of `@staticmethod` seems unrelated to the fix. I think it should go in a separate commit.

File: mesonbuild/linkers/detect.py [hunk 4]
```diff
@@ -71,7 +72,10 @@ def guess_win_linker(env: 'Environment', compiler: T.List[str], comp_class: T.Ty
         elif not invoked_directly:
             return linkers.ClangClDynamicLinker(
                 for_machine, override, exelist=compiler, prefix=comp_class.LINKER_PREFIX,
-                version=search_version(o), direct=False, machine=None)
+                version=search_version(o), direct=False, machine=None,
+                rsp_syntax=RSPFileSyntax.MSVC
+                if (invoked_directly or comp_class.get_argument_syntax() == 'msvc')
```
Comment (bruchar1) [inline mesonbuild/linkers/detect.py:77]: You are inside a `if not invoked_directly`. Therefore, you don't need to test for `invoked_directly` here...

File: mesonbuild/compilers/asm.py (hunk 3 above)
Comment (kasper93) [reply mesonbuild/compilers/asm.py:160]: done

File: mesonbuild/linkers/detect.py (hunk 4 above)
Comment (kasper93) [reply mesonbuild/linkers/detect.py:77]: done

**Review by bruchar1 — APPROVED** [review]
LGTM

**Comment by benoit-pierre** [comment]
Related #12252?

**Comment by kasper93** [comment]
> Related #12252?

Doesn't look like it is. In this case response file is used, but is generated with wrong escape sequence for given compiler/linker invocation.

**Comment by kasper93** [comment]
Bump. Good time to merge after release.

**Comment by thesamesam** [comment]
Give it a rebase for a fresh CI run?

File: mesonbuild/linkers/detect.py [hunk 5]
```diff
@@ -71,7 +72,10 @@ def guess_win_linker(env: 'Environment', compiler: T.List[str], comp_class: T.Ty
         elif not invoked_directly:
             return linkers.ClangClDynamicLinker(
                 for_machine, override, exelist=compiler, prefix=comp_class.LINKER_PREFIX,
-                version=search_version(o), direct=False, machine=None)
+                version=search_version(o), direct=False, machine=None,
+                rsp_syntax=RSPFileSyntax.MSVC
+                if comp_class.get_argument_syntax() == 'msvc'
+                else RSPFileSyntax.GCC)
```
Comment (eli-schwartz) [inline mesonbuild/linkers/detect.py:78]: I find this line wrapping hard to read.

Comment (kasper93) [reply mesonbuild/linkers/detect.py:78]: The wrapping style is required by linter. I didn't have much choice.

Comment (eli-schwartz) [reply mesonbuild/linkers/detect.py:78]: I'm pretty sure our linter configuration doesn't include rules such as requiring that the `x if y else z` construct be split across three lines.

In general I'm fairly skeptical at the idea of linters that "require" you to do things a certain way. The job of a linter is usually to detect aberrant patterns and tell you not to use them, not to be prescriptive about what you do to solve it. That is e.g. why meson does not (and never will) use `black`

Comment (kasper93) [reply mesonbuild/linkers/detect.py]: > I'm pretty sure our linter configuration doesn't include rules such as requiring that the x if y else z construct be split across three lines.

I separated it into two lines, if you prefer that. I can make one line too. Tell me what's best for you.

Comment (eli-schwartz) [reply mesonbuild/linkers/detect.py]: We can probably just calculate the value once at the top of the function, since we already know the comp_class and whether it is invoked_directly.

Comment (kasper93) [reply mesonbuild/linkers/detect.py]: done

**Comment by kasper93** [comment]
Rebased again to latest master.

**Comment by jon-turney** [comment]
> There is `get_argument_syntax()` and `rsp_file_syntax()` which ultimately means exactly the same thing

As an aside, this seems like dangerous reasoning. They are different attributes of a tool (the style of command line switches and the style of escaping used in any rsp file, respectively).  There might be only two archetypes presently ('msvc'-like and 'gcc'-like), but making these "the same thing" seems like a bad design decision.
